pybamm/solvers/c_solvers/klu.cpp

```diff
@@ -0,0 +1,429 @@
+using jacobian_type = std::function<py::array_t<double>(double, py::array_t<double>, double)>;
+using event_type = std::function<py::array_t<double>(double, py::array_t<double>)>;
+using np_array = py::array_t<double>;
+
+using jac_get_type = std::function<np_array()>;
+
+#pragma GCC visibility push(hidden)
+class PybammFunctions
+{
+
+public:
+    int number_of_states;
+    int number_of_events;
+
+    PybammFunctions(const residual_type &res,
+                    const jacobian_type &jac,
+                    const jac_get_type &get_jac_data_in,
+                    const jac_get_type &get_jac_row_vals_in,
+                    const jac_get_type &get_jac_col_ptrs_in,
+                    const event_type &event,
+                    const int n_s,
+                    int n_e)
+        : number_of_states(n_s),
+          number_of_events(n_e),
+          py_res(res),
+          py_jac(jac),
+          py_event(event),
+          py_get_jac_data(get_jac_data_in),
+          py_get_jac_row_vals(get_jac_row_vals_in),
+          py_get_jac_col_ptrs(get_jac_col_ptrs_in)
+    {
+    }
+
+    py::array_t<double> operator()(double t, py::array_t<double> y, py::array_t<double> yp)
+    {
+        return py_res(t, y, yp);
+    }
+
+    py::array_t<double> res(double t, py::array_t<double> y, py::array_t<double> yp)
+    {
+        return py_res(t, y, yp);
+    }
+
+    void jac(double t, py::array_t<double> y, double cj)
+    {
+        // this function evaluates the jacobian and sets it to be the attribute of a python class
+        // which can then be called by get_jac_data, get_jac_col_ptr, etc
+        py_jac(t, y, cj);
+    }
+
+    np_array get_jac_data()
+    {
+        return py_get_jac_data();
+    }
+
+    np_array get_jac_row_vals()
+    {
+        return py_get_jac_row_vals();
+    }
+
+    np_array get_jac_col_ptrs()
+    {
+        return py_get_jac_col_ptrs();
+    }
+
+    np_array events(double t, np_array y)
+    {
+        return py_event(t, y);
+    }
+
+private:
+    residual_type py_res;
+    jacobian_type py_jac;
+    event_type py_event;
+    jac_get_type py_get_jac_data;
+    jac_get_type py_get_jac_row_vals;
+    jac_get_type py_get_jac_col_ptrs;
+};
+#pragma GCC visibility pop
+
+int residual(realtype tres, N_Vector yy, N_Vector yp, N_Vector rr, void *user_data)
+{
+    PybammFunctions *python_functions_ptr = static_cast<PybammFunctions *>(user_data);
+    PybammFunctions python_functions = *python_functions_ptr;
+
+    realtype *yval, *ypval, *rval;
+    yval = N_VGetArrayPointer(yy);
+    ypval = N_VGetArrayPointer(yp);
+    rval = N_VGetArrayPointer(rr);
+
+    int n = python_functions.number_of_states;
+    py::array_t<double> y_np = py::array_t<double>(n, yval);
+    py::array_t<double> yp_np = py::array_t<double>(n, ypval);
+
+    py::array_t<double> r_np;
+
+    r_np = python_functions.res(tres, y_np, yp_np);
+
+    double *r_np_ptr = (double *)r_np.request().ptr;
```

Comment:
Note: you can use:
```
    auto r = x.unchecked<1>(); // x must have ndim = 1; can be non-writeable
```
to get access to the array without bounds checking etc, which is faster, see here: https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html#direct-access

The docs say that this directly references the data, so you might be able to get a raw pointer from this to pass to sundials without copying..?